### PR TITLE
Updated map management

### PR DESCRIPTION
**Pull request by Mak-Sym**
Hazelcast Management Center allows to [re-configure maps at runtime](http://docs.hazelcast.org/docs/management-center/3.9-EA/manual/html/Managing_Maps.html). This is very useful feature, but it allows to tweak store configuration for distributed maps only, whereas near cache store config remains untouched. There are many business cases, when near cache for distributed maps needs to be updated as well, and as for now, there is no way to re-configure the near cache without cluster re-start.

This PR is draft implementation (proof of concept), which enables near cache configuration tweaks at runtime (similar to what is done for distributed maps).  Management Center needs to be updated separately.

What's done in scope of this PR:
1. Updated `UpdateMapConfigOperation` to apply near cache config changes
2. Updated near cache implementation to use config suppliers instead of final config objects, for "dynamic" updates of any config changes.

### CODE REVIEW

**Comment by Donnerbart** [comment]
Hmm, after a quick peek the PR looks technically correct, but I have a bad gut feeling to insert `getService().getMapServiceContext().getMapContainer(getName()).getMapConfig().getNearCacheConfig()` calls in any fastpath of the Near Cache. I'll try to dig out my Near Cache benchmarks to see if it will have a noticeable performance effect.

I think an alternative could be to re-create the Near Cache itself, once a config has been changed. We could re-use the existing record store, so no cached data is lost or has to be copied. Just the "wrapper" aka `DefeaultNearCache` will be replaced, with the new config values, which could remain normal or final fields. This could also make the PR smaller, since a lot of stuff doesn't have to be changed. We would just need another "smart" constructor which assembles a Near Cache with a new config and the existing record store.

**Comment by Donnerbart** [comment]
I have to revise my opinion about the "technical correct". If we change the key serialization or in-memory format of a Near Cache, this will break now. There is another Near Cache implementation for `NATIVE` in-memory format, which would need to be used, when switching from `OBJECT` to `BINARY` (even if we throw away all data in the Near Cache etc.). Also if you add a Near Cache to a data structure, which had no Near Cache before, we need to re-create the proxy classes. This will be a lot of work, especially since you cannot work on the EE parts.

We could limit the config changes to existing Near Caches and no changes in serialization formats. Then you could still change the TTL, `maxIdleSeconds`, `cacheLocalEntries` and `localUpdatePolicy`. I'm not 100% sure about the `EvictionConfig`, since we have additional EE policies. Also the `NearCachePreloaderConfig` could be difficult to change. It's the question if this is worth the effort and performance regression risk.

**Comment by Mak-Sym** [comment]
Thank you for the feedback @Donnerbart , it makes perfect sense. `UpdateMapConfigOperation` allows to change only some limited settings for distributed maps, and the same approach I wanted to use for NearCached maps.

> which would need to be used, when switching from OBJECT to BINARY
I would not allow to change memory format. Only allow to change fields, which are "changeable" for distributed maps

> Also if you add a Near Cache to a data structure
I also wouldn't do it. From mine draft implementation:
`if (oldConfig.isNearCacheEnabled() && mapConfig.isNearCacheEnabled()) {`

The main idea of these changes is to allow to change same settings for near cached maps as are allowed to change for distributed maps (by using Hazelcast management center)

And the business case to address you concerns related reasoning for such changes (all names and dates are changed :)): Customer X uses near cached distributed maps to cache the data. Because of the bug in deployment pipeline, max limit for all the caches was set to `Integer.MAX_VALUE` (both near cache size and distributed cache size). After some time, caches started consuming all the memory. As the cluster uptime is critical for the Company X business, they considered using Hazelcast Management Center to change cache sizes without cluster restart. But the problem is that Management Center allows to change only size of the distributed part of the map, leaving near cache limits unchanged. At the end of the day, they had to restart the cluster (after fixing the deployment issue)

A workaround could be to monitor cache sizes and clear them manually as their size is bigger than some threshold before the problem is fixed, but clearly it doesn't work and scale very well.

**Comment by Mak-Sym** [comment]
Another idea to save performance would be to create `initEvictor()` method in `NearCache` (similar to how it's done in `MapContainer`) to re-init evictors on config change. If that's done that way, then potentially slow calls like `getService().getMapServiceContext().getMapContainer(getName()).getMapConfig().getNearCacheConfig()` are no longer needed. Also it will be clear that changes are limited to eviction config only (no memory format changes etc.)

Will be happy to work on this if it makes sense for you, @Donnerbart

**Comment by Holmistr** [comment]
@mmedenjak Any plans for this PR?

**Comment by ahmetmircik** [comment]
I see this PR is addressing some possible use cases. But doing near cache config update from management center may cause confusion and unexpected results.

Currently, we have map config update from management center and it does partial map config update for some shared map config parameters of cluster members. But Near cache is a bit different than imap since it is an application local feature and it is legitimate to see different near cache configs on different cluster members or clients.

Maybe better to think a new way to address this issue...

**Comment by Holmistr** [comment]
@Mak-Sym @ahmetmircik I'm going to close this PR, since there is no real activity, sorry :(
@Mak-Sym If you were willing to continue on this, please comment and we can resume where we ended. However, we really appreciate your effort that you tried to contribute. Very sorry it ended like this. Please, comment in case you're still eager to continue :)

**Comment by Mak-Sym** [comment]
@Holmistr thank you for heads up. I still think re-configuration of near cached maps the same way distributed maps can be configured (in other words - bringing feature parity between distributed and near cached maps) is important. I'm happy to finish this work if Hazelcast team is interested in this fix.

**Comment by Holmistr** [comment]
Hey @Mak-Sym , glad to see that you're still with us :) Let me circle back to our team and I'll let you know.

**Comment by ahmetmircik** [comment]
Hi @Mak-Sym, first let me rephrase the issue, while there is a single map config per map, there can be many near-cache configs for the same map(every node and client can have its own near-cache config). 
For now, i have no proposal for the solution but if you come up with one, i will be happy to merge it.
